### ISSUE TITLE
Adguard breaks normal browsing on macOS with Eset firewall enabled

### ISSUE THREAD

**Issue by tomx420** [AdguardTeam/AdguardForMac#698]
### **Issue:**
Cannot browse normally with Adguard turned ON. 
Normal browsing using any browser Firefox, Chrome or Safari will not work and returns error such as "Safari can't find the server" or "Site can't be reached". 
Issue occurs on macOS (10.14 and above) with Adguard installed and third party firewall enabled. Internet security suite used is Eset Cybersecurity Pro. 

### Steps to reproduce
1. Install macOS 10.14 or above
2. Install Adguard latest version (or nightly)
3. Install 3rd party firewall - Eset Cybersecurity 
4. Use any browser - Firefox, Chrome or Safari

### Attempted solutions
1. Updated software to latest version
2. tried using nightly version
3. Removed cert check
4. Reinstalling Certificates

### Actual behavior
For Adguard to work with this setup, it will require either of the following:
1. Adguard requests firewall to be turned OFF or manually create a firewall rule: "Allow any inbound connection from the internet on any port" (usually on high ports 40,000-65535).

****or**** 

2. If Third party firewall is completely removed, the  macOS built in firewall will require a rule for "Allow incoming connections" on any port for any internet address. 

### **See Images:**

<details><summary>Screenshot:</summary>

**Eset firewall rule:** 
![image](https://user-images.githubusercontent.com/61563431/80293811-34e58780-87a6-11ea-8488-cdb7701bd2cb.png)

**MacOS built-in Firewall rule:**
![image](https://user-images.githubusercontent.com/61563431/80293910-32376200-87a7-11ea-9932-fa0104c37d69.png)

**Logged firewall rule when enabled:** 
![image](https://user-images.githubusercontent.com/61563431/80294197-3022d280-87aa-11ea-883c-6d62150f63d0.png)

</details>

Turning OFF firewall or creating inbound rules will create additional security risks to the user's workstation and would be better of not using Adguard. 

**P.S.:** 
Any users using Adguard on mac, recommend using NMAP to check if there are incoming ports open with Adguard installed. 
The results will show some high ports are open and listening when a port scan is done. 

### Expected behavior
Normal browsing without turning  off firewall or creating security vulnerabilities by opening inbound connections on firewall. 
Browser should open website as normal. 

### Feature update requested:
Requesting feature update to allow app to work without interfering with host machine's firewall.

### Customer ID 
Issue - 2506038
Multiple logs, screenshots and port scan results were sent to @Adguard support team, but response is: 
1. this issue is inherent with macOS 
2. and recommendation to turn OFF firewall. 

Turning OFF firewall or creating any internet inbound rules will create additional security risks to the user's workstation and would be better of not using Adguard. 

### Your environment
OS: MacOS Catalina 10.15.4 (19E287)

Browser: 
Safari - Version 13.1 (15609.1.20.111.8)
Firefox - 75.0 (64-bit)
Chrome - Version 81.0.4044.122 (Official Build) (64-bit)

AV/Internet Security: Eset Cybersecurity Pro 6.8.300.0

Adguard versions (tried): 
Adguard for Mac - version 2.3.2 from December 23, 2019
Adguard Nightly - version 2.4.7.777

Any specific potentially conflicting software installed: (e.g. antiviruses, firewalls, traffic counters, cleaners) - None.

Also refer similar issue:
#695

**Comment by ameshkov** [comment]
Is it possible in Eset to create a more sophisticated rule? What’s really needed is a rule that allows accepting inbound connections from localhost, not the entire network

**Comment by tomx420** [comment]
> Is it possible in Eset to create a more sophisticated rule? What’s really needed is a rule that allows accepting inbound connections from localhost, not the entire network

Yes tried putting a rule with local IP address but it still wouldn't work. Also doesnt explain why the macOS in-built firewall is also requiring a similar rule when Eset was removed.

**Comment by ameshkov** [comment]
AdGuard itself works as a local proxy.

Here's what kernel extension does for every new connection:
1. Checks if the connection belongs to an app from the list of filtered apps
2. If it does, the kext reroutes it to the local proxy that is run by AG.
3. Then AG on behalf of the app creates the real remote connection.

So I don't really understand why would we need that exception. I've even tried removing that exception and checking "Block all incoming connections" in the Firewall settings" and I haven't observed any issues.

@Stillness-2 any idea? Do we do that for the case when the user enables "use AG as a proxy"?

@tomx420 what for Eset, it's hard to say what's wrong. I can say what we need for sure:

1. Inbound connections from localhost (either 127.0.0.1 or ::)
2. AdGuard should be able to create any outbound connection (as it operates on behalf of the filtered app/browser)

**Comment by tomx420** [comment]
@ameshkov , Thanks for the response.
I created a rule to allow all inbound from localhost in Eset. Also I have made an allow inbound & outbound for Adguard on all ports too. 
And it still didnt work; the web browser was stuck and unable to resolve.

See attached images
![image](https://user-images.githubusercontent.com/61563431/80580399-d1ca4e00-8a4e-11ea-8006-2df14da91234.png)

![image](https://user-images.githubusercontent.com/61563431/80580457-e575b480-8a4e-11ea-89a6-55a7a6e4e9af.png)

Not sure why Adguard requires an inbound connection from the internet specifically  to function on the Eset firewall and on dynamic high port ranges. 

I used to have this working fine till Dec 2019; Some update in early Jan 2020 has created the issue.

**Comment by tomx420** [comment]
> AdGuard itself works as a local proxy.
> 
> Here's what kernel extension does for every new connection:
> 
> 1. Checks if the connection belongs to an app from the list of filtered apps
> 2. If it does, the kext reroutes it to the local proxy that is run by AG.
> 3. Then AG on behalf of the app creates the real remote connection.
> 
> So I don't really understand why would we need that exception. I've even tried removing that exception and checking "Block all incoming connections" in the Firewall settings" and I haven't observed any issues.
> 
> @Stillness-2 any idea? Do we do that for the case when the user enables "use AG as a proxy"?
> 
> @tomx420 what for Eset, it's hard to say what's wrong. I can say what we need for sure:
> 
> 1. Inbound connections from localhost (either 127.0.0.1 or ::)
> 2. AdGuard should be able to create any outbound connection (as it operates on behalf of the filtered app/browser)

@ameshkov , thanks for the reply. 
Eset is actually filtering the localhost connections and terminating Adguard local proxy connections. 

Now, Eset Internet Security for Windows already has in-built fw rule that allows "All traffic within the computer" which allows all internal local host communication. 

So, as suggested and as shown in previous comment, I added a rule in Mac Eset to allow all localhost connection by using both 127.0.0.1 and ::1 but that didn't work either. 
For some reason Eset in Mac has some issue with IPv4/IPv6 and after doing some research I found that I had to use ::0 for the fw rule to work. 


**### Suggested Resolution:** 
1. Add firewall rule for allow inbound connections from localhost (127.0.0.1 or ::1).
For Eset Cybersecurity Pro (Mac) - Use the following Firewall Rule: 
All Application - Inbound - TCP - All Remote Ports - All Local Ports - ::0 (the unspecified ipv6 address)
2. Allow Adguard to make an outbound connection on all ports to the internet.
Adguard - Outbound - TCP&UDP - All Remote Ports - All Local Ports - Internet (All internet destination)


Eset Rule should like this:
![image](https://user-images.githubusercontent.com/61563431/80769399-66889500-8b90-11ea-90ae-92083af3fa86.png)


I had sent numerous emails to your tech support but none of them suggested what you asked for. So once again @ameshkov , thanks for the reply, it helped me look where I had to, to get this fixed. 
Great Product and love this app; keep up the great work. 
Much appreciate!!
Cheers!!
